README.md
# [Tag Cloud](https://github.com/code-guru/tag-cloud)



### How to Install

```sh
$ npm install tag-cloud
```

### Getting Started

Start the server with the following command and open the link http://localhost:3000

```sh
$ npm start
```

### How to Test

Run one, or a combination of the following commands to lint and test your code:

```sh
$ npm run lint          # Lint the source code
$ npm test              # Run unit and systems tests
$ npm run test:unit     # Run unit tests
$ npm run test:system   # Run system tests
$ npm run test:watch    # Run unit tests and watch files for changes
$ npm run test:cover    # Run unit tests with code coverage by Istanbul
```

### Test Coverage
```
  Tag Cloud system tests
    API
      ✓ /api/twitter?q=amazing (252ms)
      ✓ /api/feed?q=feedurl (67ms)
      ✓ /api/feed?q=nonValidURL
      ✓ /api/feed?q=
    Homepage api
      ✓ return with html
      ✓ return with 404 on non-existing pages

  Tag Cloud unit tests
    apiRoutes.js
      ✓ should returns a router object
    expressApp.js
      ✓ should returns an express app object
    feedClient.js
      ✓ should be able to fetch a feed
    twitterClient.js
      ✓ should be able to search twitter for a string
    tagTransformers.js
      ✓ should be able to get stream of twits, and transform it to tags
      ✓ should be able to get stream of feeds and transform it to tags
    utils.js
      ✓ should be able to convert html to text
      ✓ should be able to detect stop words


  14 passing (424ms)


=============================== Coverage summary ===============================
Statements   : 94.57% ( 122/129 )
Branches     : 65% ( 13/20 )
Functions    : 33.33% ( 1/3 )
Lines        : 94.26% ( 115/122 )
================================================================================
```

### Screenshots
![Twitter](http://i.imgur.com/C9TQ0J1.png)
![Feed](http://i.imgur.com/7y5Ipj6.png)

### License

MIT © 2016 Mehdi Tirgar


CHANGELOG.md
## yojstest Change Log

All notable changes to this project will be documented in this file.

### [Unreleased][unreleased]

- ...

### [v1.0.0] - 2016-01-01

- ...

[unreleased]: https://github.com/code-guru/yojstest/compare/v1.0.0...HEAD
[v0.0.1]: https://github.com/code-guru/yojstest/compare/v0.0.0...v1.0.0


tests/unit/unitTest.js
/**
 * TagCloud
 *
 * Copyright © 2016 Mehdi Tirgar. All rights reserved.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE.txt file in the root directory of this source tree.
 */

'use strict';

const expect = require('chai').expect;
const _ = require('highland');
const nock = require('nock');
const path = require('path');

describe('Tag Cloud unit tests', function () {
  describe('apiRoutes.js', function () {
    const apiRoute = require('../../src/apiRoutes');
    it('should returns a router object', function () {
      expect(apiRoute.name).to.be.eql('router');
      expect(apiRoute.stack).to.have.length.of.at.least(3);
    });
  });

  describe('expressApp.js', function () {
    const expressApp = require('../../src/expressApp');
    it('should returns an express app object', function () {
      expect(expressApp.get('env')).to.be.eql(expressApp.settings.env)
    });
  });

  describe('feedClient.js', function () {
    const fetchFeed = require('../../src/feedClient');
    const transformer = (query) => stream => stream;
    before(() => nock.load(path.resolve(__dirname, '../mocks/feed.nock.json')));
    it('should be able to fetch a feed', function (done) {
      fetchFeed('https://news.ycombinator.com/rss', transformer).then((result) => {
        expect(result).to.have.length.of(30);
      }).then(done).catch(done);
    });
  });

  describe('twitterClient.js', function () {
    const twitter = require('../../src/twitterClient');
    const transformer = (query) => stream => stream;
    before(() => nock.load(path.resolve(__dirname, '../mocks/twitter.nock.json')));
    it('should be able to search twitter for a string', function (done) {
      twitter('amazing', transformer).then((result) => {
        const statuses = result[0].data.statuses;
        expect(statuses).to.have.length.of(100);
      }).then(done).catch(done);
    });
  });

  describe('tagTransformers.js', function () {
    const transformers = require('../../src/tagTransformers');
    it('should be able to get stream of twits, and transform it to tags', function (done) {
      const twits = [{
        data: {
          statuses: [
            {text: 'Lorem ipsum dolor sit amet'},
            {text: 'Nulla sit amet lorem accumsan'},
            {text: 'faucibus magna ipsum vel tellus'},
            {text: 'magna dolor porta elit'},
            {text: 'Praesent sit amet tortor non leo '}
          ]
        }
      }];
      _(twits).through(transformers.tweetsToWords('')).errors(done).collect().pull((err, result) => {
        const expectedResult = [{word: 'amet', count: 3, size: 10},
          {word: 'dolor', count: 2, size: 7},
          {word: 'ipsum', count: 2, size: 7},
          {word: 'lorem', count: 2, size: 7},
          {word: 'magna', count: 2, size: 7},
          {word: 'sit', count: 3, size: 10}];
        expect(result).to.be.eql(expectedResult);
        done();
      })
    });

    it('should be able to get stream of feeds and transform it to tags', function (done) {
      const feed = [
        {title: 'Integer molestie leo tellus'},
        {title: 'eu nisi condimentum molestie'},
        {title: 'tortor non leo pellentesque'},
        {title: 'Duis commodo consectetur tellus'},
        {title: 'metus risus finibus tellus'}
      ];
      _(feed).through(transformers.feedToWords()).errors(done).collect().pull((err, result) => {
        const expectedResult = [{word: 'leo', count: 2, size: 7},
          {word: 'molestie', count: 2, size: 7},
          {word: 'tellus', count: 3, size: 10}];
        expect(result).to.be.eql(expectedResult);
        done();
      })
    })

  });

  describe('utils.js', function () {
    const utils = require('../../src/utils');
    it('should be able to convert html to text', function () {
      const html = '<div><span id="123">cloud</span><span class="high">la vita</div>';
      const text = utils.htmlToText(html);
      expect(text).to.be.equal('cloudla vita')
    });

    it('should be able to detect stop words', function () {
      expect(utils.isStopWord('this')).to.be.true;
      expect(utils.isStopWord('gälla')).to.be.true;
      expect(utils.isStopWord('javascript')).to.be.false;
      expect(utils.isStopWord('sverige')).to.be.false;
    })
  });

});


tests/system/systemTest.js
/**
 * TagCloud
 *
 * Copyright © 2016 Mehdi Tirgar. All rights reserved.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE.txt file in the root directory of this source tree.
 */

'use strict';

const expect = require('chai').expect;
const request = require('supertest');
const nock = require('nock');
const path = require('path');
const tagCloudApp = require('../../src/expressApp');


const requestApp = request(tagCloudApp);

describe('Tag Cloud system tests', function () {

  describe('API', function () {
    before(() => {
      nock.load(path.resolve(__dirname, '../mocks/twitter.nock.json'));
      nock.load(path.resolve(__dirname, '../mocks/feed.nock.json'));
    });

    it('/api/twitter?q=amazing', function (done) {
      const expectedResult = require('../test-data/twitter.route.expectation.json');
      requestApp.get('/api/twitter?q=amazing')
        .set('Accept', 'application/json')
        .expect(200)
        .expect(expectedResult)
        .end(done);
    });

    it('/api/feed?q=feedurl', function (done) {
      const expectedResult = require('../test-data/feed.route.expectation.json');
      requestApp.get('/api/feed?q=https://news.ycombinator.com/rss')
        .set('Accept', 'application/json')
        .expect(200)
        .expect(expectedResult)
        .end(done);
    });

    it('/api/feed?q=nonValidURL', function (done) {
      requestApp.get('/api/feed?q=nonValidURL')
        .set('Accept', 'application/json')
        .expect(500)
        .expect(function (res) {
          expect(res.body.success).to.be.false;
          expect(res.body.message).to.be.equal('The provided feed URL (nonValidURL) is not a valid web URI.');
        })
        .end(done);
    });

    it('/api/feed?q=', function (done) {
      requestApp.get('/api/feed?q=')
        .set('Accept', 'application/json')
        .expect(500)
        .expect(function (res) {
          expect(res.body.success).to.be.false;
          expect(res.body.message).to.be.equal('Query parameter is missing.');
        })
        .end(done);
    });
  });

  describe('Homepage api', function () {
    it('return with html', function (done) {
      requestApp.get('/')
        .expect('Content-Type', /html/)
        .expect(200, done);
    });

    it('return with 404 on non-existing pages', function (done) {
      requestApp.get('/hello')
        .expect('Content-Type', /html/)
        .expect(404, done);
    });

  });
});


src/utils.js
/**
 * TagCloud
 *
 * Copyright © 2016 Mehdi Tirgar. All rights reserved.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE.txt file in the root directory of this source tree.
 */

'use strict';

const fs = require('fs');
const path = require('path');
var cheerio = require('cheerio');


const stopWords = ['../stop_words/english.csv', '../stop_words/swedish.csv'].reduce((acc, filename) => {
  return acc.concat(fs.readFileSync(path.resolve(__dirname, filename), 'utf8').split(/\s/));
}, []);

const htmlToText = html => {
  const $ = cheerio.load('<body>' + html + '</body>');
  const text = $('body').text().replace(/<[^>]+>/ig, '');
  return (text || '').replace(/<style([\s\S]*?)<\/style>/gi, '')
    .replace(/<script([\s\S]*?)<\/script>/gi, '')
    .replace(/<\/div>/ig, '\n')
    .replace(/<\/li>/ig, '\n')
    .replace(/<li>/ig, '  *  ')
    .replace(/<\/ul>/ig, '\n')
    .replace(/<\/p>/ig, '\n')
    .replace(/<br\s*[\/]?>/gi, '\n')
    .replace(/<[^>]+>/ig, '');
};

module.exports = {
  htmlToText,
  isStopWord: word => stopWords.indexOf(word.toLowerCase()) >= 0
};


src/apiRoutes.js
/**
 * TagCloud
 *
 * Copyright © 2016 Mehdi Tirgar. All rights reserved.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE.txt file in the root directory of this source tree.
 */

'use strict';

const express = require('express');
const router = express.Router();
const parseFeedUrl = require('./feedClient');
const parseTwitterTag = require('./twitterClient');
const transformers = require('./tagTransformers');

router.get('*', (req, res, next) => {
  if (!req.query.q) {
    return next(new Error('Query parameter is missing.'));
  }
  return next();
});

router.get('/twitter', (req, res, next) => {
  const tag = req.query.q;
  parseTwitterTag(tag, transformers.tweetsToWords).then(res.send.bind(res)).catch(next);
});

router.get('/feed', (req, res, next) => {
  const link = req.query.q;
  parseFeedUrl(link, transformers.feedToWords).then(feedItems => {
    res.send(feedItems);
  }).catch(next);
});


router.use((err, req, res, next) => {
  res.status(err.status || 500);
  res.send({
    success: false,
    message: err.message,
    error: req.app.get('env') === 'development' ? err.stack : ''
  });
});


module.exports = router;


src/feedClient.js
/**
 * TagCloud
 *
 * Copyright © 2016 Mehdi Tirgar. All rights reserved.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE.txt file in the root directory of this source tree.
 */

'use strict';

const FeedParser = require('feedparser');
const request = require('request');
const validUrl = require('valid-url');
const debug = require('debug')('tag-cloud:feedclient');
const _ = require('highland');


module.exports = (feedUrl, transformer) => new Promise((resolve, reject) => {
  if (!validUrl.isWebUri(feedUrl)) {
    return reject(new Error(`The provided feed URL (${feedUrl}) is not a valid web URI.`));
  }

  const requestOptions = {
    url: feedUrl,
    timeout: 10000,
    gzip: true,
    headers: {
      'user-agent': 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_8_5) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/31.0.1650.63 Safari/537.36',
      accept: 'text/html,application/xhtml+xml'
    }
  };

  const parserOptions = {
    addmeta: false,
    feedurl: feedUrl
  };

  const req = request(requestOptions);
  // req.setMaxListeners(50);
  const feedparser = new FeedParser(parserOptions);

  req.on('error', onError)
    .on('response', function (res) {
      const responseStream = this;
      if (res.statusCode !== 200) {
        return this.emit('error', new Error('Bad status code'));
      }
      return responseStream.pipe(feedparser);
    });

  feedparser.on('error', onError);

  return _(feedparser)
    .errors(onError)
    .through(transformer())
    .toArray(resolve);

  function onError(err) {
    debug(err, err.stack);
    return reject(err);
  }
});


src/twitterClient.js
/**
 * TagCloud
 *
 * Copyright © 2016 Mehdi Tirgar. All rights reserved.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE.txt file in the root directory of this source tree.
 */

'use strict';

const Twit = require('twit');
const _ = require('highland');
const debug = require('debug')('tag-cloud:twitterclient');

const T = new Twit({
  consumer_key: 'XBVXg7MGym1hqH5sorSFym1PM',
  consumer_secret: 'Gq7AbbXfAH8K4y1YGUnRwyrhgIsG0CEY3YMz7Ws0peMb4BQoda',
  app_only_auth: true,
  timeout_ms: 10 * 1000
});

module.exports = (query, transformer) => new Promise((resolve, reject) => {
  _([query])
    .map(x => ['search/tweets', {q: `${x}`, count: 100}])
    .flatMap(args => _(T.get.apply(T, args)))
    .errors(onError)
    .through(transformer(query))
    .toArray(resolve);

  function onError(err) {
    debug(err, err.stack);
    reject(err);
  }
});


src/tagTransformers.js
/**
 * TagCloud
 *
 * Copyright © 2016 Mehdi Tirgar. All rights reserved.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE.txt file in the root directory of this source tree.
 */

'use strict';
const _ = require('highland');
const math = require('mathjs');
const utils = require('./utils');

const addSize = stream => stream
  .collect()
  .flatMap(words => {
    if (words.length === 0) {
      return words;
    }
    const counts = words.map(word => word.count);
    const sum = math.sum(counts);
    const max = math.max(counts);
    const factor = math.floor(10 / (max / sum));
    return _(words).map(word => Object.assign({}, word, {size: math.ceil(factor * word.count / sum)}));
  });

const toWords = stream => stream
  .reduce1((a, b) => `${a} + ${b}`)
  .map(utils.htmlToText)
  .map(x => x.replace(/https?:\/\/[^\s]*/g, ''))                   /* remove links */
  .map(x => x.replace(/[^\x00-\x7F]/g, ''))                        /* remove non-ascii characters */
  .splitBy(/[”“\-<>,\.;:_'\*\^’`´\?\+=\(\)!"€%&\/~\[\]\\\|{}\s]+/) /* break down to wors with all separators */
  .reject(x => x.length < 3)                                       /* remove all words with less than 3 characters*/
  .reject(utils.isStopWord)                                        /* remove all stop words (e.g. the, am, is, are) */
  .reject(x => x - parseFloat(x) + 1 >= 0)                         /* remove all numbers */
  .map(x => x.toLowerCase());                                      /* convert words to lowercase */

const commonTransform = stream => stream
  .group(x=>x)                                                     /* group same words in the list together */
  .flatMap(_.pairs)                                                /* create key value pairs */
  .map(x => ({word: x[0], count: x[1].length}))                    /* create collection of { word, count } objects */
  .reject(x => x.count < 2)                                        /* remove all words with less than 2 occurrence */
  .sortBy((a, b) => b.count - a.count)                             /* sort by number of occurrence */
  .take(50)                                                        /* limit result to 50 */
  .sortBy((a, b) => a.word.replace(/[@#]/, '').localeCompare(b.word.replace(/[@#]/, ''), 'se-SV'))
  .through(addSize);                                               /* add size parameter */


module.exports = {
  feedToWords: () => stream =>
    stream
      .pick(['title', 'summary', 'description'])
      .flatMap(_.values)
      .reject(x => typeof x !== 'string')
      .through(toWords)
      .through(commonTransform),
  tweetsToWords: (query) => stream =>
    stream
      .flatMap(x => _(x.data.statuses))
      .pluck('text')
      .through(toWords)
      .reject(word => word === query)
      .reject(word => word.replace(/[#@]/, '') === query.replace(/#@/, ''))
      .through(commonTransform)
};



assets/javascripts/app.js
'use strict';
var $ = window.$ || window.jQuery;

$(function () {
  var submitBtn = $('.cloud-container button');
  $('form').on('submit', function (e) {
    var activePane = $('.tab-pane.active');
    var type = activePane.attr('id');
    var query = activePane.find('[type="text"]').val();
    e.preventDefault();
    if (!type || !query) { return; }
    submitBtn.attr('disabled', true).text('Loading...');
    displayTags(type, query);
  });

  function displayTags(type, query) {
    var tagToHtml = function (tag) {
      return '<span class="word tag' + tag.size + '"><a href="#tag" title=' + tag.count + '"times">' + tag.word + '</a></span>';
    };
    var request = $.ajax({
      url: '/api/' + type,
      method: 'GET',
      data: {q: query},
      dataType: 'json'
    });

    request.done(function (tags) {
      var tagsHtml = tags.map(tagToHtml);
      $('#' + type + ' .tags').removeClass('hidden').html(tagsHtml);
    });

    request.fail(function (jqXHR, textStatus) {
      var message = textStatus;
      try {
        message = JSON.parse(jqXHR.responseText).message;
      } catch (e) {
        console.error(jqXHR.responseText);
      }
      $('#error-alert').text('Request failed: ' + message).removeClass('hidden').show();
    });

    request.always(function () {
      submitBtn.attr('disabled', false).text('Go!');
    });
  }
});


src/expressApp.js
/**
 * TagCloud
 *
 * Copyright © 2016 Mehdi Tirgar. All rights reserved.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE.txt file in the root directory of this source tree.
 */

'use strict';

const express = require('express');
const path = require('path');
const favicon = require('serve-favicon');
const logger = require('morgan');
const cookieParser = require('cookie-parser');
const bodyParser = require('body-parser');

const apiRoutes = require('./apiRoutes');

const app = express();

app.set('views', path.join(__dirname, '../views'));
app.set('view engine', 'ejs');

if (app.get('env') === 'development') {
  app.use(logger('dev'));
}

app.use(bodyParser.json());
app.use(bodyParser.urlencoded({extended: false}));
app.use(cookieParser());
app.use(express.static(path.join(__dirname, '../assets')));

app.use('/api', apiRoutes);

app.get('/', (req, res, next) => {
  res.render('index', {title: 'Tag Cloud'});
});

app.use((req, res, next) => {
  const err = new Error('Not Found');
  err.status = 404;
  next(err);
});

app.use((err, req, res, next) => {
  res.status(err.status || 500);
  res.render('error', {
    message: err.message,
    error: app.get('env') === 'development' ? err : {}
  });
});


module.exports = app;
